src/Propulsion.SqlStreamStore/LockedIngester.fs

```diff
@@ -0,0 +1,229 @@
+            | None     -> new BlockingCollection<InternalBatch>(ConcurrentQueue())
+
+        let stats = Stats(logger, ?statsInterval = statsInterval)
+
+        let commit (handle: IDisposable) batch =
+            async {
+                try
+                    try
+                        do! ledger.CommitPosition { Stream = streamId; ConsumerGroup = consumerGroup; Position = Nullable(batch.lastPosition) }
+                        stats.UpdateCommitedPosition(batch.lastPosition)
+                        logger.Debug("Committed position {position}", batch.lastPosition)
+
+                        // Handle end of stream if reached.
+                        if batch.isEnd then
+                            do onCaughtUp ()
+                    with
+                    | exc ->
+                        logger.Error(exc, "Exception while commiting position {position}", batch.lastPosition)
+                        return! Async.Raise exc
+                finally
+                    // Release the lock after committing the position.
+                    logger.Debug("Releasing the lock after commiting")
+                    do handle.Dispose()
+            }
+
+        member this.SubmitBatch (batch: InternalBatch) =
+            async {
+                let! ct = Async.CancellationToken
+                do work.Add(batch, ct)
+                stats.UpdateQueue(work)
+            }
+
+        member private this.TakeBatch () =
+            async {
+                let! ct = Async.CancellationToken
+                let batch = work.Take(ct)
+                stats.UpdateQueue(work)
+                return batch
+            }
+
+        member this.Start () =
+            async {
+                // Start reporting stats
+                do! Async.StartChild stats.Start |> Async.Ignore
+
+                let mutable workItem = Work.Take
+
+                let! ct = Async.CancellationToken
+                while not ct.IsCancellationRequested do
+                    try
+                        // Take work if haven't done it already or finished previous batch
+                        let! batch =
+                            async {
+                                match workItem with
+                                | Work.Batch b -> return b
+                                | Work.Take ->
+                                    let! b = this.TakeBatch()
+                                    workItem <- Work.Batch b
+                                    return b
+                            }
+
+                        stats.UpdateBatch(batch)
+
+                        logger.Debug("Trying to obtain the lock")
+
+                        // Try obtaining a distributed sql lock for the consumer group.
+                        // The lock would be released as part of commit operation on the underlying ingester.
+                        let! handle = ledger.Lock(consumerGroup)
```

Comment:
Locking is to allow multiple instances of a projection service running for a single consumer group for redundancy. I'm quite aware this is a bottle neck, but I'm not sure we can do any better without partitioning anyway?

Reply:
For better or worse, Propulsion.EventStore assumes a single instance globally...
You can then use an orchestrator to manage the instance count; the writer is then expected to act idempotently in the face of multiple instances etc.

Reply:
(discussed out of band - the locking is not intrinsic to the functionality, the `proProjector` template will illustrate its use)